### PR TITLE
Mongo custom expressions

### LINKED ISSUE #7097: MongoDB support Custom Column and Custom Expression
### Feature requests and proposals

It seems that MongoDB "Custom" questions do not offer support for "Custom Expressions". 

Here's what I see when I use my PostgreSQL database:

screenshot from 2018-03-08 18-52-51

And these are the options for MongoDB:

screenshot from 2018-03-08 18-53-14

Since MongoDB's aggregation pipeline is essentially a JavaScript function, I believe it should be possible to include custom expressions for this database as well. Maybe even allowing arbitrary expressions instead of just `+`, `-`, `*`, `/`.

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

### PR DESCRIPTION
**Pull request by howonlee**
We only have the ticket #9382 which is for bigquery, but we're apparently doing both according to notion product tracker so here is the beginning of the mongo custom expressions.

~~This handles very many but not all of the needed custom expressions but this portion is the most straightforward part. What's missing is:~~
- ~~median~~
- ~~percentile~~
- ~~variance~~
- ~~distinct~~
- ~~coalesce~~
~~These require futzing with the aggregation step more and are coming in a separate PR~~

- ~~case~~
- ~~share~~
- ~~sum-where~~
- ~~count-where~~
~~These require futzing with the case statements in mongo which seem fiddly in my testing on local mongo and will come in separate PR also.~~

~~I think median and percentile are not feasible and distinct is taken care of in a way I didn't understand previously. I don't know about variance. Coalesce is being implemented~~

Jeff was confused by the mongo pipeline docs and I was too, a majority of the above is already implemented...

### CODE REVIEW

**Comment by howonlee** [comment]
It looks like a fair majority of the work is gonna be mongo as opposed to BQ because mongo is different than all the SQL dealies

File: modules/drivers/mongo/test/metabase/driver/mongo/query_processor_test.clj [hunk 1]
```diff
@@ -153,6 +153,101 @@
...
+                 "category_id" "$category_id",
+                 "bob" {"$abs" "$latitude"},
+                 "cobb" {"$ceil" {"$abs" "$latitude"}},
+                 "longitude" "$longitude",
+                 "latitude" "$latitude",
+                 "price" "$price"}}
+               {"$limit" 5}],
+              :collection "venues",
+              :mbql? true}
+             (qp/query->native
```
Comment (dpsutton) [inline modules/drivers/mongo/test/metabase/driver/mongo/query_processor_test.clj:244]: I think these tests could look a little nicer if you had a nice helper function and it was called like
```clojure
(is (= {"bob"  {"$abs" "$latitude"},
        "cobb" {"$ceil" {"$abs" "$latitude"}}}
       (some-test-fn {:bob   [:abs (mt/id :latitude)]
                      :cobb  [:ceil [:expression "bob"]]})))
```

**Comment by dpsutton** [comment]
LGTM. Left note on tests to make them highlight the changes we care about. One failed test is the maria change which is fixed on master. Do you know what's going on with the others?

File: modules/drivers/mongo/test/metabase/driver/mongo/query_processor_test.clj (hunk 1 above)
Comment (howonlee) [reply modules/drivers/mongo/test/metabase/driver/mongo/query_processor_test.clj:244]: not 100% a fan of terseness in testing but I guess making it terse will make it conform better to rest of file so sgtm

**Comment by howonlee** [comment]
Looks like extremely strange flakes

**Comment by howonlee** [comment]
Now that I actually look harder at end-user interface again, I note that we don't need separate points of view on the aggregator-dealies since we don't seem to offer it? So it's just case and coalesce, lol

**Comment by howonlee** [comment]
cam has informed me that the way to whack the differential version thing is by implementing #13136. I think introspecting is possible but introspecting for version is kinda exploding the scope right now so just add in a version

File: modules/drivers/mongo/src/metabase/driver/mongo.clj [hunk 2]
```diff
@@ -18,6 +18,7 @@
             [monger.collection :as mc]
             [monger.command :as cmd]
             [monger.conversion :as m.conversion]
+            [monger.core :as mcore]
```
Comment (dpsutton) [inline modules/drivers/mongo/src/metabase/driver/mongo.clj:21]: nit but we already have `[monger.core :as mg]` in two other places and should continue this practice

File: modules/drivers/mongo/src/metabase/driver/mongo.clj [hunk 3]
```diff
@@ -174,8 +175,10 @@
 (defmethod driver/describe-database :mongo
   [_ database]
   (with-mongo-connection [^com.mongodb.DB conn database]
-    {:tables (set (for [collection (disj (mdb/get-collection-names conn) "system.indexes")]
-                    {:schema nil, :name collection}))}))
+    {:tables  (set (for [collection (disj (mdb/get-collection-names conn) "system.indexes")]
+                    {:schema nil, :name collection}))
+     :version (:version (mcore/command conn {:buildInfo 1}))
+     }))
```
Comment (dpsutton) [inline modules/drivers/mongo/src/metabase/driver/mongo.clj:181]: nit these brackets should be on the line above

**Comment by dpsutton** [comment]
@howonlee this code doesn't compile:

> Unable to resolve symbol: mapv->rvalue in this context

I cannot even require the namespace `metabase.driver.mongo` because you have `(str/split "\.")`. And `str/split` requires a regex not a string in the second position (it is in a thread macro here) and `"\."` is an unreadable string

```clojure
Clojure 1.10.3
user=> "\."
Syntax error reading source at (REPL:1:4).
Unsupported escape character: \.
```

If you prepend a `#` to get a regex it is valid, but then the entire form uses `feature` which is undefined here
```clojure
(defmethod driver/supports? [:mongo feature] [_ _ [db]] (= "5" (-> (get-in db [:details :version])
                                                                   (str/split #"\.")
                                                                   (first))))
```

So this PR feels like perhaps it is still a draft and not ready to be reviewed. Further, I think the `driver/supports?` usage with adding a db like you did here is a non-starter unfortunately.

**Comment by howonlee** [comment]
> @howonlee this code doesn't compile:
> 
> > Unable to resolve symbol: mapv->rvalue in this context
> 
> I cannot even require the namespace `metabase.driver.mongo` because you have `(str/split "\.")`. And `str/split` requires a regex not a string in the second position (it is in a thread macro here) and `"\."` is an unreadable string
> 
> ```clojure
> Clojure 1.10.3
> user=> "\."
> Syntax error reading source at (REPL:1:4).
> Unsupported escape character: \.
> ```
> 
> If you prepend a `#` to get a regex it is valid, but then the entire form uses `feature` which is undefined here
> 
> ```clojure
> (defmethod driver/supports? [:mongo feature] [_ _ [db]] (= "5" (-> (get-in db [:details :version])
>                                                                    (str/split #"\.")
>                                                                    (first))))
> ```
> 
> So this PR feels like perhaps it is still a draft and not ready to be reviewed. Further, I think the `driver/supports?` usage with adding a db like you did here is a non-starter unfortunately.

It was ready and then I decided to add the syncing in this branch, yes, sorry

File: src/metabase/sync/sync_metadata/tables.clj [hunk 4]
```diff
@@ -80,6 +81,15 @@
 
 ;;; ---------------------------------------------------- Syncing -----------------------------------------------------
 
+(s/defn ^:private update-database-metadata!
+  "If there is a version in the db-metadata update the DB to have that in the DB model"
+  [database :- i/DatabaseInstance db-metadata :- i/DatabaseMetadata]
+  (log/info (trs "Found new version for DB:")
```
Comment (jeff303) [inline src/metabase/sync/sync_metadata/tables.clj:87]: Does this add a space between the message and the param value?  I haven't seen this usage of `trs` before (only seen using, like `{0}` and then making the val appear as an additional param after the message).

File: src/metabase/sync/sync_metadata/tables.clj [hunk 5]
```diff
@@ -137,10 +147,10 @@
                         :description description))))
 
 
-(s/defn ^:private db-metadata :- #{i/DatabaseMetadataTable}
-  "Return information about `database` by calling its driver's implementation of `describe-database`."
-  [database :- i/DatabaseInstance]
-  (set (for [table (:tables (fetch-metadata/db-metadata database))
+(s/defn ^:private table-set :- #{i/DatabaseMetadataTable}
+  "Get set of tables from database metadata which aren't metabase metadata tables"
```
Comment (jeff303) [inline src/metabase/sync/sync_metadata/tables.clj:151]: I find this comment confusing.  Any way to reword it?

File: modules/drivers/mongo/test/metabase/driver/mongo_test.clj [hunk 6]
```diff
@@ -96,11 +119,11 @@
 
 (deftest describe-database-test
   (mt/test-driver :mongo
-    (is (= {:tables #{{:schema nil, :name "checkins"}
-                      {:schema nil, :name "categories"}
-                      {:schema nil, :name "users"}
-                      {:schema nil, :name "venues"}}}
-           (driver/describe-database :mongo (mt/db))))))
+    (is (= #{{:schema nil, :name "checkins"}
+            {:schema nil, :name "categories"}
```
Comment (jeff303) [inline modules/drivers/mongo/test/metabase/driver/mongo_test.clj:123]: nit: I think all these lines need one extra space of indentation

**Review by jeff303 — COMMENTED** [review]
Mostly nits but LGTM otherwise

**Review by jeff303 — APPROVED** [review]

**Pull request merged**